### PR TITLE
CB2-7573 Handle changes to test result test types and defects

### PR DESCRIPTION
**Pull request by Ian-Hodges**
## Description
**CB2-7573:** When a test is modified and reprocessed into NOP, it can result in the old and new result being attached to the record. This PR adds code to delete results if they already exist prior to processing the updated record.
**CB2-7685:** The test_result table has a trigger that populates the testtype_version table. It has been found that it can affect the identity returned to the update store lambda. This PR has changes to address this.

Related issue: [CB2-7573](https://dvsa.atlassian.net/browse/CB2-7573)
Related issue: [CB2-7685](https://dvsa.atlassian.net/browse/CB2-7685)

### Clean pipeline build
image
### Successful VOTT regression pack
image
### Performance Check
To check the effect of the deletion steps on the performance of the NOP Rebuild, a test results run was completed. While the run was in progress the lambda was flipped between delete and upsert mode to see if there were any noticeable changes in performance.

#### Lambda Duration
image
*Red sections are where deletions were enabled.

#### MySql Activity
image
*Red sections are where deletions were enabled.

As can be seen there is no discernible difference between delete and upsert modes.
## CB2-7685
The test_result table has a trigger that populates the testtype_version table. It has been found that it can affect the identity returned to the update store lambda.

The issue only arises when:

A result is processed for the second time and results in an update rather than an insert.
and
The test_result does not differ in any anyway to the record being updated.

### Evidence of completion
In the absence of a fix or understanding of the behaviour, a workaround has been implemented which involves inserting a timestamp along with the rest of the test_result data.

Before workaround
First upsert
image
Second upsert
image

After workaround
First upsert
image
Second upsert
image
Data in the database
image

### CODE REVIEW

File: src/services/test-result-record-conversion.ts [hunk 1]
```diff
@@ -122,13 +129,28 @@ const upsertTestResults = async (testResults: TestResults): Promise<void> => {
...
                         lastUpdatedById,
+                        moment().format("YYYY-MM-DD HH:mm:ss.SSS"),
                     ],
                     testResultConnection
                 );
 
                 continue;
             }
 
+            if (!process.env.DISABLE_DELETE_ON_UPDATE) {
```
Comment (ChrisPeska-BJSS) [inline src/services/test-result-record-conversion.ts:140]: This really nice to have. Do you think the feature toggle should be a covered by testing?

**Review by ChrisPeska-BJSS — DISMISSED** [review]
Great work @Ian-Hodges. The change grew out it scope marginally but I think it does look better this way. 

Looking at the change I realised about there is a potential risk lurking in a wider NOP design. Queues are not FIFO so if VTM produces consecutive changes within a fairly small timespan, there is a chance the messages will be processed in opposite order to the changes made. (concurrency is tricky to comprehend)

I left one question, which might be addressed later if it is relevant one.

File: src/models/tech-record.ts [hunk 2]
```diff
@@ -210,7 +210,7 @@ const parseTechRecord = (image: DynamoDbImage): TechRecord => {
         grossKerbWeight: image.getNumber("grossKerbWeight"),
         grossLadenWeight: image.getNumber("grossLadenWeight"),
         grossGbWeight: image.getNumber("grossGbWeight"),
-        grossDesignWeight: image.getNumber("grossGbWeight"),
+        grossDesignWeight: image.getNumber("grossDesignWeight"),
```
[the remark this answers is not in the document]
Comment (Ian-Hodges) [reply src/models/tech-record.ts:213]: Wasn't me :-) I think it was the MOTH team who found it.

**Comment by Ian-Hodges** [comment]
> Looking at the change I realised about there is a potential risk lurking in a wider NOP design. Queues are not FIFO so if VTM produces consecutive changes within a fairly small timespan, there is a chance the messages will be processed in opposite order to the changes made. (concurrency is tricky to comprehend)

I'd thought the same. I had planned to check how long it takes for a dynamo event to reach NOP. I'll raise a ticket so it doesn't get missed.

File: src/services/test-result-record-conversion.ts (hunk 1 above)
Comment (Ian-Hodges) [reply src/services/test-result-record-conversion.ts:140]: Done. It was tricky to get both working! 😃
Ran through the pipeline:
image

**Review by naathanbrown — APPROVED** [review]

**Review by ChrisPeska-BJSS — APPROVED** [review]
All good 👍

**Pull request merged**